### PR TITLE
Tool to get image digest

### PR DESCRIPTION
**Pull request by ash2k**
Used in https://github.com/bazelbuild/rules_docker/pull/445.

### CODE REVIEW

**Comment by ash2k** [comment]
@mattmoor could you take a look please?

**Comment by ittaiz** [comment]
@matmoor any news?

**Comment by ixdy** [comment]
@dekkagaijin can you take a look?

**Comment by ittaiz** [comment]
@mattmoor any news?

**Comment by ittaiz** [comment]
@ash2k the build is failing btw

**Comment by dekkagaijin** [comment]
@deft-code @KaylaNguyen PTAL

**Comment by ash2k** [comment]
@ittaiz the build got stuck because of docker login it seems. I cannot do anything about it, someone needs to push a button to restart it.

**Comment by ittaiz** [comment]
I think you can close and then reopen the PR and it will re-trigger.
Another option is an empty commit.

File: tools/image_digest_.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from containerregistry.client.v2_2 import oci_compat
+from containerregistry.tools import logging_setup
+
+from six.moves import zip  # pylint: disable=redefined-builtin
+
+
+parser = argparse.ArgumentParser(
+    description='Calculate digest for a container image.')
+
+parser.add_argument(
```
Comment (KaylaNguyen) [inline tools/image_digest_.py]: Please specify types and whether is required for these flag values.

File: tools/image_digest_.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  config = args.config
+  if args.config:
+    logging.info('Reading config from %r', args.config)
+    with open(args.config, 'r') as reader:
+      config = reader.read()
+  elif args.tarball:
+    logging.info('Reading config from tarball %r', args.tarball)
+    with v2_2_image.FromTarball(args.tarball) as base:
+      config = base.config_file()
+
```
Comment (KaylaNguyen) [inline tools/image_digest_.py]: Is tarball guaranteed to exists? What would happen if neither config and tarball exists?

File: tools/image_digest_.py (hunk 1 above)
Comment (ash2k) [reply tools/image_digest_.py]: They are all strings, except for the `--oci` which is boolean and `action='store_true'` ensures it is either true or false. Is that ok or I need to explicitly put all types there? Please note that I've just copied [fast_pusher_.py](https://github.com/atlassian/containerregistry/blob/0492daa2652d500ea618bebe66cdf2e3a80d030c/tools/fast_pusher_.py) and removed what is not needed, following the style as close as possible. It does not have types or `required` markers.

File: tools/image_digest_.py (hunk 2 above)
Comment (ash2k) [reply tools/image_digest_.py]: There is/was a check above that terminated the program if neither config nor tarball parameters are provided. I moved it here into an `else` branch. Reads better now. (this is originally copied from another file too).

**Comment by ash2k** [comment]
Travis blocks on docker login each time. 🤷‍♂️

File: tools/image_digest_.py (hunk 1 above)
Comment (KaylaNguyen) [reply tools/image_digest_.py]: It's up to you at this point. I'm happy with what you have now. You don't _have_ to follow fast_pusher_.py code style.

**Comment by KaylaNguyen** [comment]
Can you try what @ittaiz mentioned to see if that fixes the docker login problem?
Restarting travis build didn't help.

**Comment by dekkagaijin** [comment]
The root cause of the `docker login` hang is that travis won't divulge the credentials unless the change lives in a branch of `github.com/google/containerregistry`. @KaylaNguyen if you want to test it on github, you can clone this PR's repo (atlassian/containerregistry:image_digest) and push it to a new branch of google/containerregistry, which will allow the tests to run.

**Comment by KaylaNguyen** [comment]
I'm reluctant to test and merge this PR on github, as internal google code is currently the source of truth and changing code externally would make internal and external container registry code diverge.
As @jonjohnsonjr suggested, one way we can make this work is to check this code into our internal code base, go through the review process and integration tests, then manually export and create a release to github.
What do you think @ash2k?

**Comment by ash2k** [comment]
@KaylaNguyen I'm happy with anything as long as this code will get into this repo.

**Comment by KaylaNguyen** [comment]
Resolving :)

**Comment by ash2k** [comment]
@KaylaNguyen any updates?

**Comment by KaylaNguyen** [comment]
I'm working on it. Out for review, pending adding test.

**Comment by KaylaNguyen** [comment]
The change was rolled out. I'm closing this PR :)

**Comment by ash2k** [comment]
@KaylaNguyen thanks a lot!

**Comment by ittaiz** [comment]
Awesome!

**Comment by ash2k** [comment]
Would it be possible to cut a release so that I can use it in my rules_docker PR bazelbuild/rules_docker#445 ? Thanks!

**Comment by jonjohnsonjr** [comment]
@ash2k Ack. There are two small changes I want to land before this release, but I'll try to cut it today 👍

**Comment by jonjohnsonjr** [comment]
@ash2k done now (thanks @KaylaNguyen!)

**Pull request closed without merging**